File: tools/demo_registry.py
#!/usr/bin/env python3
"""Shared browser-demo registry for the demo audit tools.

Keep demo identity, base paths, default ports, and lightweight API markers in one
place so static docs checks, local HTTP smoke tests, and hosted-link probes fail
together when a browser demo moves.
"""
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path


VALID_LIFECYCLE_STATES: tuple[str, ...] = (
    "hosted-keep",
    "promote-next",
    "local-keep",
    "archive",
    "tombstone",
)


@dataclass(frozen=True)
class Demo:
    name: str
    base_path: str
    api_path: str
    page_marker: str
    extra_args: tuple[str, ...] = ()
    default_port: int = 0
    hosted_path: str = ""
    hosted_port: int = 0
    hosted_api_keys: tuple[str, ...] = ()


@dataclass(frozen=True)
class LifecycleDecision:
    state: str
    issue: int
    reason: str


DEMOS: tuple[Demo, ...] = (
    Demo("guarddemo", "/guarddemo", "api/scenarios", "safety floor", ("-jobs", "1"), 8151),
    Demo(
        "turntaxdemo", "/turntax", "api/suites", "turn-tax demo", ("-jobs", "1"), 8150,
        hosted_path="/", hosted_port=8150, hosted_api_keys=("suites",),
    ),
    Demo(
        "ctxdemo", "/ctxdemo", "api/scenarios", "multi-agent context demo", ("-jobs", "1"), 8153,
        hosted_path="/", hosted_port=8153, hosted_api_keys=("models", "scenarios"),
    ),
    Demo(
        "demorace", "/demorace", "api/ladder", "reuse demo", ("-jobs", "1"), 8147,
        hosted_path="/demorace/", hosted_api_keys=("models", "prefill_tok_ratio_a_over_c"),
    ),
    Demo("dropindemo", "/dropin", "api/gallery", "fak guard drop-in gallery", default_port=8154),
    Demo("unseedemo", "/unsee", "api/events", "Un-See It", default_port=8156),
    Demo("timewolfdemo", "/timewolf", "api/scenarios", "what time is it, Mr. Wolf?", default_port=8155),
    Demo("trychatdemo", "/trychat", "api/suggestions", "try-it agentic chat", default_port=8157),
)


DEMO_LIFECYCLE: dict[str, LifecycleDecision] = {
    "guarddemo": LifecycleDecision(
        "local-keep",
        1738,
        "healthy security-floor proof; keep in local/deep catalog unless promoted by a later rubric pass",
    ),
    "turntaxdemo": LifecycleDecision(
        "hosted-keep",
        1167,
        "hosted front-door efficiency proof; live link currently witnessed",
    ),
    "ctxdemo": LifecycleDecision(
        "hosted-keep",
        1739,
        "hosted research/performance slot; must re-earn with current model and net-value witness",
    ),
    "demorace": LifecycleDecision(
        "hosted-keep",
        1739,
        "hosted live-model research slot; must re-earn with current model and net-value witness",
    ),
    "dropindemo": LifecycleDecision(
        "local-keep",
        1738,
        "healthy adoption proof; keep in local/deep catalog unless it earns a front-door slot",
    ),
    "unseedemo": LifecycleDecision(
        "local-keep",
        1738,
        "healthy KV-removal witness; keep in local/deep catalog unless promoted by the rubric",
    ),
    "timewolfdemo": LifecycleDecision(
        "promote-next",
        1736,
        "healthy LCD agentic loop; next candidate for hosted agentic card",
    ),
    "trychatdemo": LifecycleDecision(
        "promote-next",
        1736,
        "healthy LCD try-it chat; next candidate for hosted agentic card",
    ),
}


def demo_map(demos: tuple[Demo, ...] = DEMOS) -> dict[str, Demo]:
    return {d.name: d for d in demos}


def lifecycle_map(
    decisions: dict[str, LifecycleDecision] = DEMO_LIFECYCLE,
) -> dict[str, LifecycleDecision]:
    return dict(decisions)


def lifecycle_defects(
    demos: tuple[Demo, ...] = DEMOS,
    decisions: dict[str, LifecycleDecision] = DEMO_LIFECYCLE,
) -> list[str]:
    registered = {d.name: d for d in demos}
    decision_names = set(decisions)
    defects: list[str] = []
    for name in sorted(set(registered) - decision_names):
        defects.append(f"browser demo lacks lifecycle decision: cmd/{name}")
    for name in sorted(decision_names - set(registered)):
        defects.append(f"lifecycle decision names unknown browser demo: cmd/{name}")
    for name in sorted(set(registered) & decision_names):
        decision = decisions[name]
        demo = registered[name]
        if decision.state not in VALID_LIFECYCLE_STATES:
            defects.append(
                f"{name}: invalid lifecycle state {decision.state!r}; "
                f"want one of {', '.join(VALID_LIFECYCLE_STATES)}"
            )
        if decision.issue <= 0:
            defects.append(f"{name}: lifecycle decision missing GitHub issue number")
        if not decision.reason.strip():
            defects.append(f"{name}: lifecycle decision missing reason")
        if demo.hosted_path and decision.state != "hosted-keep":
            defects.append(f"{name}: hosted demo must have lifecycle hosted-keep, got {decision.state}")
        if not demo.hosted_path and decision.state == "hosted-keep":
            defects.append(f"{name}: lifecycle hosted-keep requires hosted_path metadata")
    return defects


def discover_browser_demo_names(workspace: Path) -> list[str]:
    cmd_dir = workspace / "cmd"
    return sorted(p.parent.name for p in cmd_dir.glob("*/page.html"))


def registry_defects(workspace: Path, demos: tuple[Demo, ...] = DEMOS) -> list[str]:
    discovered = set(discover_browser_demo_names(workspace))
    registered = {d.name for d in demos}
    defects: list[str] = []
    for name in sorted(discovered - registered):
        defects.append(f"browser demo has page.html but is not in DEMOS registry: cmd/{name}")
    for name in sorted(registered - discovered):
        defects.append(f"DEMOS registry names a browser demo without page.html: cmd/{name}")
    return defects


def select_demos(names: list[str] | None, demos: tuple[Demo, ...] = DEMOS) -> tuple[list[Demo], list[str]]:
    if not names:
        return list(demos), []
    by_name = demo_map(demos)
    unknown = [name for name in names if name not in by_name]
    return [by_name[name] for name in names if name in by_name], unknown


def normalize_base(base_path: str) -> str:
    p = "/" + base_path.strip("/")
    return "" if p == "/" else p


def hosted_url(host: str, demo: Demo) -> str:
    if not demo.hosted_path:
        return ""
    port = f":{demo.hosted_port}" if demo.hosted_port else ""
    path = "/" if normalize_base(demo.hosted_path) == "" else f"{normalize_base(demo.hosted_path)}/"
    return f"http://{host}{port}{path}"


def hosted_demo_urls(host: str, demos: tuple[Demo, ...] = DEMOS) -> dict[str, str]:
    return {demo.name: url for demo in demos if (url := hosted_url(host, demo))}


def demo_url(port: int, base_path: str, rel: str = "") -> str:
    base = normalize_base(base_path)
    suffix = (rel or "").lstrip("/")
    if suffix:
        return f"http://127.0.0.1:{port}{base}/{suffix}"
    return f"http://127.0.0.1:{port}{base}/"

File: internal/hooks/demo_registry.go
package hooks

import (
	"sort"
	"strings"
)

// demo_registry.go — the Go twin of tools/demo_registry.py, the shared browser-demo registry the
// demo audit tools key on (demo identity, base paths, default ports, hosted metadata) plus the
// lifecycle decisions. It backs the DEMO_COMMAND coverage gate (demoBrowserNames, derived here) and
// the BROWSER_CONTRACT gate (gate_browsercontract.go), exactly as demo_registry.py backs both
// demo_command_audit.py and demo_browser_contract.py.
//
// Parity anchor: tools/demo_registry.py DEMOS / DEMO_LIFECYCLE / VALID_LIFECYCLE_STATES /
// lifecycle_defects(). gate_browsercontract_test.go pins this table against the live Python
// demo_registry so a registry add/remove/edit that skips this file reds `go test`.

// demoReg is one browser-demo registry entry — the Go twin of demo_registry.Demo. Only the fields
// the ported audits read are carried; the smoke-test-only extra_args are omitted.
type demoReg struct {
	name          string
	basePath      string
	apiPath       string
	pageMarker    string
	defaultPort   int
	hostedPath    string
	hostedPort    int
	hostedAPIKeys []string
}

// demoRegistry mirrors demo_registry.DEMOS verbatim, in the same order (the order coverage and
// contract defects are emitted in).
var demoRegistry = []demoReg{
	{name: "guarddemo", basePath: "/guarddemo", apiPath: "api/scenarios", pageMarker: "safety floor", defaultPort: 8151},
	{name: "turntaxdemo", basePath: "/turntax", apiPath: "api/suites", pageMarker: "turn-tax demo", defaultPort: 8150,
		hostedPath: "/", hostedPort: 8150, hostedAPIKeys: []string{"suites"}},
	{name: "ctxdemo", basePath: "/ctxdemo", apiPath: "api/scenarios", pageMarker: "multi-agent context demo", defaultPort: 8153,
		hostedPath: "/", hostedPort: 8153, hostedAPIKeys: []string{"models", "scenarios"}},
	{name: "demorace", basePath: "/demorace", apiPath: "api/ladder", pageMarker: "reuse demo", defaultPort: 8147,
		hostedPath: "/demorace/", hostedAPIKeys: []string{"models", "prefill_tok_ratio_a_over_c"}},
	{name: "dropindemo", basePath: "/dropin", apiPath: "api/gallery", pageMarker: "fak guard drop-in gallery", defaultPort: 8154},
	{name: "unseedemo", basePath: "/unsee", apiPath: "api/events", pageMarker: "Un-See It", defaultPort: 8156},
	{name: "timewolfdemo", basePath: "/timewolf", apiPath: "api/scenarios", pageMarker: "what time is it, Mr. Wolf?", defaultPort: 8155},
	{name: "trychatdemo", basePath: "/trychat", apiPath: "api/suggestions", pageMarker: "try-it agentic chat", defaultPort: 8157},
}

// lifecycleDecision is the Go twin of demo_registry.LifecycleDecision.
type lifecycleDecision struct {
	state  string
	issue  int
	reason string
}

// validLifecycleStates mirrors demo_registry.VALID_LIFECYCLE_STATES.
var validLifecycleStates = []string{"hosted-keep", "promote-next", "local-keep", "archive", "tombstone"}

// demoLifecycle mirrors demo_registry.DEMO_LIFECYCLE.
var demoLifecycle = map[string]lifecycleDecision{
	"guarddemo":    {"local-keep", 1738, "healthy security-floor proof; keep in local/deep catalog unless promoted by a later rubric pass"},
	"turntaxdemo":  {"hosted-keep", 1167, "hosted front-door efficiency proof; live link currently witnessed"},
	"ctxdemo":      {"hosted-keep", 1739, "hosted research/performance slot; must re-earn with current model and net-value witness"},
	"demorace":     {"hosted-keep", 1739, "hosted live-model research slot; must re-earn with current model and net-value witness"},
	"dropindemo":   {"local-keep", 1738, "healthy adoption proof; keep in local/deep catalog unless it earns a front-door slot"},
	"unseedemo":    {"local-keep", 1738, "healthy KV-removal witness; keep in local/deep catalog unless promoted by the rubric"},
	"timewolfdemo": {"promote-next", 1736, "healthy LCD agentic loop; next candidate for hosted agentic card"},
	"trychatdemo":  {"promote-next", 1736, "healthy LCD try-it chat; next candidate for hosted agentic card"},
}

// demoRegNames returns the registry demo names in registry (DEMOS) order.
func demoRegNames() []string {
	names := make([]string, len(demoRegistry))
	for i, d := range demoRegistry {
		names[i] = d.name
	}
	return names
}

// isValidLifecycleState reports whether state is one of validLifecycleStates.
func isValidLifecycleState(state string) bool {
	for _, s := range validLifecycleStates {
		if s == state {
			return true
		}
	}
	return false
}

// demoLifecycleDefects ports demo_registry.lifecycle_defects(): every registered demo needs a
// lifecycle decision (and vice versa), each decision needs a valid state, a positive issue, and a
// non-blank reason, and hosted_path <-> hosted-keep must agree in both directions. The emission
// order mirrors the Python: missing-decision names, then unknown-demo names, then the per-demo
// checks over the intersection — each block sorted by name.
func demoLifecycleDefects() []string {
	registered := map[string]demoReg{}
	for _, d := range demoRegistry {
		registered[d.name] = d
	}

	var missingDecision, unknownDemo, both []string
	for name := range registered {
		if _, ok := demoLifecycle[name]; !ok {
			missingDecision = append(missingDecision, name)
		} else {
			both = append(both, name)
		}
	}
	for name := range demoLifecycle {
		if _, ok := registered[name]; !ok {
			unknownDemo = append(unknownDemo, name)
		}
	}
	sort.Strings(missingDecision)
	sort.Strings(unknownDemo)
	sort.Strings(both)

	var defects []string
	for _, name := range missingDecision {
		defects = append(defects, "browser demo lacks lifecycle decision: cmd/"+name)
	}
	for _, name := range unknownDemo {
		defects = append(defects, "lifecycle decision names unknown browser demo: cmd/"+name)
	}
	for _, name := range both {
		decision := demoLifecycle[name]
		demo := registered[name]
		if !isValidLifecycleState(decision.state) {
			defects = append(defects, name+": invalid lifecycle state '"+decision.state+"'; want one of "+
				joinComma(validLifecycleStates))
		}
		if decision.issue <= 0 {
			defects = append(defects, name+": lifecycle decision missing GitHub issue number")
		}
		if strings.TrimSpace(decision.reason) == "" {
			defects = append(defects, name+": lifecycle decision missing reason")
		}
		if demo.hostedPath != "" && decision.state != "hosted-keep" {
			defects = append(defects, name+": hosted demo must have lifecycle hosted-keep, got "+decision.state)
		}
		if demo.hostedPath == "" && decision.state == "hosted-keep" {
			defects = append(defects, name+": lifecycle hosted-keep requires hosted_path metadata")
		}
	}
	return defects
}

// joinComma renders a slice as ", "-joined text (the Python `', '.join(...)`).
func joinComma(xs []string) string {
	out := ""
	for i, x := range xs {
		if i > 0 {
			out += ", "
		}
		out += x
	}
	return out
}
